checks.go

```diff
@@ -0,0 +1,252 @@
+// checks.go contains checks that are identical for both Linux and Windows.
+
+package main
+
+import (
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"reflect"
+	"regexp"
+	"strings"
+)
+
+// check is the smallest unit that can show up on a scoring report. It holds all
+// the conditions for a check, and its message and points (autogenerated or
+// otherwise).
+type check struct {
+	Message      string
+	Points       int
+	Fail         []cond
+	Pass         []cond
+	PassOverride []cond
+}
+
+// cond, or condition, is the parameters for a given test within a check.
+type cond struct {
+	Type    string
+	Path    string
+	Cmd     string
+	User    string
+	Group   string
+	Name    string
+	Key     string
+	Value   string
+	After   string
+}
+
+func (c cond) requireArgs(args ...interface{}) {
+	// Don't process internal calls -- assume the developers know what they're
+	// doing. This also prevents extra errors being printed when they don't pass
+	// required arguments.
+	if c.Type == "" {
+		return
+	}
+
+	v := reflect.ValueOf(c)
+	vType := v.Type()
+	for i := 0; i < v.NumField(); i++ {
+		if vType.Field(i).Name == "Type" {
+			continue
+		}
+
+		required := false
+		for _, a := range args {
+			if vType.Field(i).Name == a {
+				required = true
+				break
+			}
+		}
+
+		if required {
+			if v.Field(i).String() == "" {
+				fail(c.Type + ":", "missing required argument '" + vType.Field(i).Name + "'")
+			}
+		} else if v.Field(i).String() != "" {
+			warn(c.Type + ":", "specifying unnecessary argument '" + vType.Field(i).Name + "'")
+		}
+
+	}
+}
+
+func (c cond) String() string {
+	output := ""
+	v := reflect.ValueOf(c)
+	typeOfS := v.Type()
+
+	for i := 0; i < v.NumField(); i++ {
+		if v.Field(i).String() == "" {
+			continue
+		}
+		output += fmt.Sprintf("\t%s: %v\n", typeOfS.Field(i).Name, v.Field(i).String())
+	}
+	return output
+}
+
+func handleReflectPanic(condFunc string) {
+	if r := recover(); r != nil {
+		fail("Check type does not exist: "+condFunc, "("+r.(*reflect.ValueError).Error()+")")
+	}
+}
+
+// runCheck executes a single condition check.
+func runCheck(cond cond) bool {
+	if err := deobfuscateCond(&cond); err != nil {
+		fail(err.Error())
+	}
+	defer obfuscateCond(&cond)
+	debug("Running condition:\n", cond)
+
+	condFunc := ""
+	negation := false
+	condEnding := cond.Type[len(cond.Type)-3 : len(cond.Type)]
+	if condEnding == "Not" {
+		negation = true
+		condFunc = cond.Type[:len(cond.Type)-3]
+	} else {
+		condFunc = cond.Type
+	}
+
+	// Catch panic if check type doesn't exist
+	defer handleReflectPanic(condFunc)
+
+	// Using reflection to find the correct function to call.
+	vals := reflect.ValueOf(cond).MethodByName(condFunc).Call([]reflect.Value{})
+	result := vals[0].Bool()
+	err := vals[1]
+
+	if negation {
+		debug("Result is", !result, "(was", result, "before negation) and error is", err)
+		return err.IsNil() && !result
```

Comment:
is `err` being an `interface{}` the reason for using `IsNil` a method on `interface{}`? Weird that you can't directly use comparison operators on `interface{}` when comparing to `nil` because `interface`s are just pointers(?) I wonder what the internal implementation of `IsNil` is lol

Reply:
Reflection is weird, it returns a reflection value type instead of an interface. Documentation for that is here: https://godocs.io/reflect#Value.IsNil